### ISSUE TITLE
Transaction atomicity question

### ISSUE THREAD

**Issue by alg** [eventide-examples/funds-transfer-component#1]
Hi there.

Thanks for sharing this example. It's been very helpful for understanding architecture of services.

I've been struggling to understand the following scenario which may potentially happen due to concurrent nature of events. How do we handle the case when we already withdrew funds from one account only to discover that while we did that the destination account got closed?

My prior experience tells operations to withdraw and deposit should be a part of the same transaction and we either post all Withdrew + Deposited + Transfered events or only TransferFailed with the details of what went wrong. Here I see it broken down into the very logical sequence of events, but with a flaw I described.

I know the scenario is contrived, but it illustrates the idea. We need atomicity of transactions. What's the idiomatic solution for this in fully autonomous services?

**Comment by sbellware** [comment]
@alg In a distributed system like this, the use of locks (due to atomic transactions, etc) isn't feasible. You'd need another distributed system to control the locks, and it would become increasingly centralized, which would defeat the purpose and benefits of avoiding a monolithic system based on atomic transactions.

Locking transactions, like atomic relational database transactions can only be safely used within a single service, which is why they work fine in a monolith. In fact, it's one of the defining characteristics of a monolith. But if attempted in a distributed system, it's "turtles all the way down" in terms of building coordination layers for coordination layers.

The account component (see the account component example project for the details) processes all operations for a single account on a single machine, and processes those operations sequentially. The architecture is predicate on not processing the events of a single account on numerous machines in parallel.

That's a reality of all such systems, not just those built on something like Eventide. In fact, it would be considered a mistake to try to process events for an account in parallel in separate processes. In fact, there are countermeasures in the handler implementation code that will cause a service to crash if conditions are detected that indicate that account transactions for an account are being processed in parallel.

Systems like this are scaled through parallel processing, but that doesn't mean that any given account's transactions are distributed in parallel over a number of machines. There'd be no way to ensure that messages are processed in the order they were written. Two (or more) machines (or processes) that are processing an input category stream will process a fixed set of accounts. Any given account is only ever processed on the same machine, which ensures sequential processing, and allows protection against concurrency.

So, let's say we have two accounts: `account-123` and `account-456`. And let's say that we're running a group of two consumers of account messages: `consumer-A` and `consumer-B`. The messages for `account-123` are always processed by the same consumer, say `consumer-A`. And the messages for `account-456` are always processed by `consumer-B`. 

In truth, both of these accounts may end up being processed by the same consumer (it depends on a cryptographic hash calculated from the stream name), but they will always be processed by that exact same consumer. Any given account will never have its messages processed by different consumers, and that's a necessary and beneficial constraint of the architectural style.

This is a _partitioning_ of streams that allows for safe horizontal scaling. When talking about parallelizing, the subject of partitioning _also_ has to be addressed. Parallelization without the need for partitioning is extremely rare in practice.

**Comment by alg** [comment]
Thanks for laying it all out for me, Scott. Much appreciated. I found the section in the documentation you were referring to and studying the groups now. Giving up on the idea of transactions and locking is painful at times I must admit. :)

**Comment by sbellware** [comment]
@alg Indeed! Atomic transactions make that part of things super easy - but only that part. When we're working with atomic database transactions, it's the easiest architectural style we'll ever work with. But it's also the style that never stops getting worse over time. Working with autonomous components and streams is harder to do at the start, and it doesn't reflect any of the things we learned earlier in life, but it ends up being far more sustainable in the long term.
